### ISSUE TITLE
Implement multipart parsing

### ISSUE THREAD

**Issue by ronag** [nodejs/undici#974]
Our `extractBody` method is missing support for parsing `multipart/form-data` responses.

**Comment by ronag** [comment]
I'm not sure who/why someone would send multipart/form-data from a server so would appreciate some requests on this. Doesn't seem to be a very useful feature.

**Comment by jimmywarting** [comment]
I think it seems useful, i mean https://www.npmjs.com/package/form-data has 4.5M download / week

**Comment by MrWaip** [comment]
Hello. Telegram bot api accept only FormData for files
https://core.telegram.org/bots/api#inputfile

**Comment by blittle** [comment]
> I'm not sure who/why someone would send multipart/form-data from a server so would appreciate some requests on this. Doesn't seem to be a very useful feature.

Correct me if I'm wrong, but isn't the bigger use case the server parsing multipart form data from the client? The example from the node-fetch PR iml is:

```js
import { Response } from 'node-fetch'

http.createServer(async function (req, res) {
  const formData = await new Response(req, { headers: req.headers }).formData()
  const file = formData.get('avatar')
  
  var arrayBuffer = await file.arrayBuffer()
  var text = await file.text()
  var whatwgReadableStream = file.stream()
});
```

It would be great if undici had the same support for multipart form data.

**Comment by jimmywarting** [comment]
the example should actually be something like this:
```js
http.createServer(async function (req, res) {
  const url = `${req.protocol}://${req.get('host')}${req.originalUrl}`
  const request new Request(url, { body: req, method: req.method, headers: req.headers })
  const formdata = request.formData()
  const file = formData.get('avatar')
  
  var arrayBuffer = await file.arrayBuffer()
  var text = await file.text()
  var whatwgReadableStream = file.stream()
});
```
But it's easier/quicker to construct a Response that don't need any url or extra options
the idea of having a formdata parsing and shipping the `body.formData()` comes from service workers
a service worker is meant to intercept the request and be able to parse the payload when needed

> I'm not sure who/why someone would send multipart/form-data from a server so would appreciate some requests on this. Doesn't seem to be a very useful feature.

It's not about responding with a multipart/form-data response like in: `fetch(url).then(res => {res.formData()})`
cuz again that would be weird if a server responded with a multipart/form-data. It's rather more about parsing incoming request to the http server which alot of folk use nodejs for. 

I have long been wishing for a body.formData but I never wanted to have that kind of functionality for getting such responses back from the server. it was b/c I wished for the server request to mimic more like a service worker request and behaving more like the service worker `Request` object where parsing incomming request with `body.formData()` is essential so you wouldn't need busboy formidable or any other library that you also need to learn how it works.

But again for a server to respond with a multipart/form-data is all that bad. it's a grate thing to use for sending multiple files back to the client. and i have used it in the web for transfering files as well. 

```js
// server.js
http.createServer(async function (req, res) {
  const formData = new FormData()
  fd.set('/uploads/logo.png', new File(...))
  fd.set('/uploads/movie-introduction.mp4', new File(...))
  const body = new Response(formData).body
  stream.readable.from(stream).pipe(res)
})

// client.js
const res = await fetch('/get-files')
const fd = res.formData()
const files = [...fd]
// use the files for something
```
This is much easier than bringing in a zip for packing and unpacking on both the server and the response if you for any reason needs to get multipe files from the server. I have done this once before myself

**Comment by ronag** [comment]
@jimmywarting: I wouldn't mind adding support for it. Would you be ok if we use port `https://github.com/node-fetch/node-fetch/blob/main/src/utils/multipart-parser.js` over to undici? @mcollina wdyt?

**Comment by jimmywarting** [comment]
Be my guest. The source is based on a quite old version of [formidable](https://github.com/node-formidable/formidable/blob/master/src/parsers/Multipart.js) that i patched quite a bit in 2017 (to get rid of some core nodejs parts like `Buffer` and `Streams`) so that it would make a friendlier browser version for [github/fetch](https://github.com/github/fetch/issues/571) polyfill but never got merged. 

also take a look into [busboy](https://github.com/mscdex/busboy/issues/266) if their new parser is any faster. i haven't tested it against node-fetch new parser

I think the owner of busboy said that it dose some more extra stuff that don't co-here with the FormData parser spec like supporting non-UTF8 encodings

---

also have a other question... would this formData() parser make it into stream consumers as well?

```js
import {formData} from 'node:stream/consumers';
const fd = await formData(iterable)
```

**Comment by blittle** [comment]
We'd like to have multipart support for Shopify Hydrogen, where our API route handlers are just functions that take undici `Request` objects. See https://github.com/Shopify/hydrogen/discussions/450

**Comment by KhafraDev** [comment]
I have a working parser that just needs tests written before a PR, although tests keep stalling tap (same issue as `client-fetch.js` which causes the current failures in the CI).

```js
const server = createServer(async (req, res) => {
    const chunks = []
    for await (const chunk of req) chunks.push(chunk)

    const resp = new Response(Buffer.concat(chunks), { headers: req.headers })
    const form = await resp.formData()
    
    const same = form.get('c')
    const hope = readFileSync('./index.html', 'utf-8')

    assert(same.name, 'index.html')
    assert(hope === await same.text())
    assert(form.get('a') === 'b')
    
    res.end('lol')
    server.close()
}).listen(0)

const form = new FormData()
form.set('a', 'b')
form.set('c', new Blob([readFileSync('./index.html')]), 'index.html')

// send POST req with form body here...
```

I'm just worried as it's less than 100 lines that there may be edge cases I've missed, although I've hopefully made it fool-proof 😅.

**Comment by mcollina** [comment]
It needs to be a streaming client - the code in busboy shows more or less how complex that is. Might be interesting to try implementing it using async iterators

**Comment by jimmywarting** [comment]
Another thing i have been thinking about doing in node-fetch when parsing FormData and encountering very large files would somehow be to offload the data into some temporary file instead of allocating everything into memory...
that's when FileLike object's would come into play... I really wish that this two got implemented:
- https://github.com/nodejs/node/issues/37338
- https://github.com/nodejs/node/issues/39015
- ...and that `blob.stream()` was a bit more stream friendly - and didn't make use of `.arrayBuffer()` in ReadableStream.start()

that is something i have on my agenda

**Comment by KhafraDev** [comment]
> It needs to be a streaming client - the code in busboy shows more or less how complex that is. Might be interesting to try implementing it using async iterators

I added the ability to use async iterators which each yield an object close to
```js
{
    headers: [Object: null prototype] {
      'Content-Disposition': 'form-data',
      name: 'c',
      filename: 'index.html',
      'Content-Type': 'application/octet-stream'
    },
    content: '... file contents...'
  }
```
although I loop over the entire buffer body before parsing so I need to combine that logic 🤔 
& I'm unsure if `Content-Disposition` and `Content-Type` are headers or should be separate?

**Comment by mcollina** [comment]
Buffering the full body is not going to work well for Node.js unfortunately, it can end up using quite a lot of memory.

**Comment by jimmywarting** [comment]
I think the `content: '... file contents...'` should be streamable as well... and it should most certainly not be a string in the case where you upload binary data

**Comment by KhafraDev** [comment]
I think I can do that :D

**Comment by Rich-Harris** [comment]
I'll add SvelteKit to the list of frameworks that would definitely benefit from this — our best workaround is to monkey-patch the undici `Request` object's `formData` method, and create a node-fetch `Request` with the raw data in order to use its `formData` method: https://github.com/sveltejs/kit/pull/5292

Obviously this is sub-optimal since it doesn't permit streaming parsing, and it adds bloat to apps, but it'll do for the time being!

**Comment by mcollina** [comment]
Essentially we need to support `await response.formData()`. Is that correct?
@KhafraDev you did a lot of work for the spec compatiblity, would you like to take this on?

I'm ok to implement this here, even if it would have some major caveats for big files.

**Comment by jimmywarting** [comment]
> I'm ok to implement this here, even if it would have some major caveats for big files.

Would it be OK to write files to a `tmp` directory and using finalizers and then remove them once it's no longer referenced? I kind of did that in this PR: https://github.com/node-fetch/fetch-blob/pull/137

this feature would kindof be blocked by [Getting Files backed up by the filesystem](https://github.com/nodejs/node/issues/37340)

**Comment by kibertoad** [comment]
Is it safe to assume availability of tmp directory in cloud environments? Immutable containers are not widely used?
Wonder if it could have a fallback or an override option.

**Comment by mcollina** [comment]
> Would it be OK to write files to a tmp directory and using finalizers and then remove them once it's no longer referenced? I kind of did that in this PR: https://github.com/node-fetch/fetch-blob/pull/137

I would prefer to _not_ implement that in here but rather provide some hook that some external module can plug in to implement that behavior. There are likely so many different ways this can be done that I do not want to support them all in this module.

One note: you want the files to be removed _anyway_ on process end, take a look at https://github.com/mcollina/on-exit-leak-free as it allows you to have a finalizer that is run on process exit.

> this feature would kindof be blocked by [Getting Files backed up by the filesystem](https://github.com/nodejs/node/issues/37340)

Indeed, that should be implemented first.

**Comment by jimmywarting** [comment]
Question: Could we depend on the [busboy](https://npmjs.com/package/busboy) package or do you wish to write something own that you have more control over and not having to worry much about any other license?

busboy do really seems to be the fastest parser out there in the NPM user land for parsing multipart data...?
it dose seem to also support more (non utf8) encoded formats as well.

**Comment by mcollina** [comment]
> Question: Could we depend on the [busboy](https://npmjs.com/package/busboy) package or do you wish to write something own that you have more control over and not having to worry much about any other license?

I'm ok to use `busboy` for this. I would prefer we keep being dependency-less but re-implementing multipart parsing from scratch would slow us too much.

**Comment by irishburlybear** [comment]
> I'm not sure who/why someone would send multipart/form-data from a server so would appreciate some requests on this. Doesn't seem to be a very useful feature.

Sveltekit needs it...

**Comment by JacobMGEvans** [comment]
Cloudflare CLI Wrangler needs it 😄

**Comment by kibertoad** [comment]
PR would be very welcome!

**Comment by mcollina** [comment]
I'm happy to do a few design calls  about this if an individual or a company is serious in implementing it.

**Comment by blittle** [comment]
Shopify Hydrogen is also a +1 on this.

**Comment by jimmywarting** [comment]
I think we can establish that quite many ppl want a formData encoder right now, so i don't think any +1 me too is necessary anymore, there is already compelling evidence that it's needed/wanted and that we should definitely implement it. 

if anything +1 are only becoming more annoying and spams everyone who is subscribed to this topic
it also becomes more annoying to scan trough this issue for comments that really matters and helps towards solving this issue. so i think unnecessary +1 comments should be deleted or hidden...

**Comment by mcollina** [comment]
What we need is somebody with time  to implement this. Can somebody from Shopify, Cloudflare, Vercel (just to name a few of the companies that +1 here) help?

**Comment by JacobMGEvans** [comment]
> What we need is somebody with time to implement this. Can somebody from Shopify, Cloudflare, Vercel (just to name a few of the companies that +1 here) help?

We will be picking this up if no one else has already, either myself or @cameron-robey

**Comment by KhafraDev** [comment]
I have a WIP implementation of a FormData parser at https://github.com/KhafraDev/undici/tree/formdata-parser

It's surprisingly small for its capabilities (basic parsing, even with mismatched chunks) and should be easy enough to maintain. It has an API that is a mix between formidable (writing chunks to the parser directly) and like busboy (using an EventEmitter).

Currently it buffers the body in one big chunk, but similarly to busboy, will be moved to its own EventEmitter. For fetch's use case, this doesn't matter as the `.formData()` algorithm is extremely inefficient anyways and any effort to optimize each part's body will be nullified when re-constructing the FormData object. Even browsers and Deno parse formdatas synchronously and in one big chunk. (On a side note, my bad attempt from earlier in this thread would actually fit our usecase as, unbeknownst to me, was how many other platforms did it).

I did write it to be used outside of `fetch`, although I have no idea if it will/should be exposed. It needs tests, header parsing, error & end events, and likely much more I can't think of off the top of my head. It's also more maintenance for undici maintainers (not just me necessarily).

Usage *currently* works along the lines of
```ts
import { FormDataParser, FormData, request } from 'undici'
import { createServer } from 'http'
import { once } from 'events'

const server = createServer(async (req, res) => {
  const parser = new FormDataParser(req.headers['content-type'])

  parser.on('header', (header) => {
    /* do something with header */
  })
  parser.on('body', (chunk) => {
    /* do something with body */
  })

  for await (const chunk of req) {
    parser.write(chunk)
  }

  res.end('got formdata')
}).listen(0)

await once(server, 'listening')

const fd = new FormData()
fd.set('key', 'and value')

const blob = new Blob([/* some data */])
fd.set('file', blob, 'readme.md')

await request(`http://localhost:${server.address().port}`, {
  method: 'POST',
  body: fd
})
```

tl;dr I made a small async FormData parser that needs a lot more work. Also busboy is a better solution for now regardless of my opinion of adding dependencies for a niche feature.

**Comment by jimmywarting** [comment]
I think i would prefer busboy, it's a very fast multipart parser that scans using streamsearch (Boyer-Moore-Horspool) that is very efficient. plus it also handles different kinds of encodings (utf8 and 7bit)

**Comment by KhafraDev** [comment]
> that is very efficient

This is irrelevant in regards to `.formData()` - it's an inefficient method which is nearly unfit for servers. Other implementations use synchronous parsing of FormData strings without chunking or regard for efficiency (see: [Deno](https://github.com/denoland/deno/blob/213d831ae3403402d55d3d084b2434c3ba8da70f/ext/fetch/21_formdata.js#L401), [Firefox](https://github.com/mozilla/gecko-dev/blob/7f3ff3f4d34e7d234da4f5f5b345d2add7f30e95/dom/base/BodyUtil.cpp#L167), and [WebKit](https://github.com/WebKit/WebKit/blob/8495e27d2576170d21efe22f91ef0ffdadc233cb/Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp#L128)). Note: I couldn't find info on how Chromium parses FormDatas, however I assume it would be the same.

Also note that the experimental/WIP FormData parser I worked on *does* use bytes to compare buffers (albeit in the form of `Buffer.prototype.indexOf` and `Buffer.prototype.includes`, rather than streamsearch -- unsure of performance implications).

**Comment by kibertoad** [comment]
> This is irrelevant in regards to .formData() - it's an inefficient method which is nearly ENTIRELY unfit for servers. 

Can you clarify this statement? Is it impossible to implement it efficiently, or doing so would require significantly more effort?

**Comment by KhafraDev** [comment]
It is impossible to implement it efficiently - at the time of writing this - for the reasons outlined [here](https://github.com/nodejs/undici/issues/974#issuecomment-1018857025). Essentially, even if you parse it asynchronously & chunk the body data, you will *still* end up with `Blob` or `File` objects that contain multiple gbs of data.

Without blobs being backed up by the filesystem, memory usage will increase greatly. I know @jimmywarting has made some effort to have this implemented in node core, but unfortunately no one has tackled it just yet.

**Comment by PodaruDragos** [comment]
so now that `undici` has `mutipart/form-data` using `busboy` is @KhafraDev answer okay ?
For example is this approach reliable for getting and witting a file to disk ? 

```
createServer(async (request, response) => {
  const chunks: Buffer[] = [];

  for await (const chunk of request) chunks.push(chunk);

  const resp = new Response(Buffer.concat(chunks), {
    headers: request.headers as HeadersInit
  });
  const form = await resp.formData();

  const file = form.get('a-file') as File;
  const fileArrayBuffer = await file.arrayBuffer();

  await writeFile(file.name, Buffer.from(fileArrayBuffer));

  response.end('done');

}).listen(3000);
```

**Comment by KhafraDev** [comment]
I would recommend using busboy or @fastify/busboy, which suits nodejs more than .formData (which I have a strong disdain for)

**Comment by jimmywarting** [comment]
i think i would have done it more like this:

```js
createServer(async (request, response) => {
  const resp = new Response(request, { headers: request.headers })
  const fd = await resp.formData()
  const file = fd.get('a-file')
  await fs.promises.writeFile(file.name, file.stream())
  response.end('done')
}).listen(3000)
```

`Response` and `writeFile` both accepts a AsyncIterator that can yield uint8arrays
So there is no need to read all chunks and concat them into a single large buffer:
```ts
  const chunks: Buffer[] = [];
  for await (const chunk of request) chunks.push(chunk);
  const resp = new Response(Buffer.concat(chunks), {
```

and likewise there is no need to read the hole file data to an arrayBuffer()
```ts
  const fileArrayBuffer = await file.arrayBuffer();
  await writeFile(file.name, Buffer.from(fileArrayBuffer));
```

i think that `res.formData()` is a grate option actually as a dependency free solution by only using web-tech. 
the only pitty is that if you upload really large file then the hole formdata you get from `res.formData()` will be living in the memory. so i also agree with @KhafraDev about the usage if the files are really big.

undici / fetch / formData, have the potential to parse the multipart upload and all the files into a temporary location and write it them to the disc

then it could populate the `FormData` with disc- based blobs
```js
// kind of doing this:
fd = new FormData()
blob = await fs.openAsBlob(path, { type })
fd.append('a-file', blob, filename)
// later resolve `res.formData()` with the FormData
return fd
```

this was something that i had in the pipeline for `node-fetch` and also the reason of why i created [createTemporaryBlob / createTemporaryFile](https://github.com/node-fetch/fetch-blob/tree/main#creating-a-temporary-file-on-the-disk) in `fetch-blob`

But i haven't gotten around at implementing this yet. 
I was planing on doing it for both `res.formData()` and also `res.blob()`
but only conditionally. 
i would start reading some chunks into memory and create some few in-memory parts and if it would grow over a certain size then i would offload the data to the disc

this would match more closely to how chrome operates when using `.formData()` and also `.blob()`

**Comment by KhafraDev** [comment]
I disagree. It's an awful solution that should have never landed in undici. I regret not blocking the PR that implemented it until we could find common ground.

- It's not spec compliant (remember this)
- It's magnitudes times slower than other spec compliant solutions[...](https://twitter.com/jarredsumner/status/1625067963656335361)
- The only requests for it to be implemented were from users of packages like svelte kit or miniflare who needed compatibility with other runtimes. This could have been easily polyfilled.

For large files, a streaming implementation doesn't matter whatsoever. It's all in memory anyways, as you can't opt-in for node to save the blobs FormData creates as file-based blobs. For small files, a streaming implementation doesn't make sense for obvious reasons.

> undici / fetch / formData, have the potential to parse the multipart upload and all the files into a temporary location and write it them to the disc

Which is easier said than done. How do we decide which blobs use the file-based solution? Is it even faster? Do we leave it up to the users, who would have to install undici anyways (and rather could have just installed @fastify/busboy), to configure themselves? How do we handle permissions, filesystem errors, cleaning up files, etc.?

> this would match more closely to how chrome operates when using .formData() and also .blob()

Chrome's .formData parser is also synchronous, as it should be according to the spec, as it is in every other browser and server environment, so comparing undici/node to any other implementation is not a good comparison. I don't see a reason why we should follow chrome with regards to how it offloads files while we are completely ignoring the spec.

**Comment by KhafraDev** [comment]
I'd also like to mention that unlike busboy or @fastify/busboy, there are no limits to file sizes, number of files/fields, or anything. There is no way to limit the size of a file, nor the number of files or fields, nor file name length, or any other limit that the aforementioned packages have. It's a giant security risk to use it in your http server.

**Comment by jimmywarting** [comment]
> How do we handle 

# Permissions
🤷‍♂️
# Filesystem errors
could fallback to in-memory based blobs
# Cleaning up files
the process of cleaning up files can be managed using finalizers, as exemplified in [this reference](https://github.com/node-fetch/fetch-blob/blob/a1a182e5978811407bef4ea1632b517567dda01f/from.js#L70C1-L80C2), losing references to the blob could delete the file.

# Decide which blobs use the file-based solution
For determining the appropriate solution for blobs, particularly in the context of res.blob() in Chrome, factors such as the content-length can play a role. If the data is not compressed and the content-length is available, a direct data pipe to the disk could be considered. However, for compressed data (gzip, br, deflate), or when there's no content-length due to streaming, an approach involving memory storage might be preferred. This could entail reading the stream into memory until a certain threshold, then transferring the remaining data to a disk-based file. Alternatively, a combination of in-memory and disk-based blobs could be used.

For res.formData(), a straightforward streaming process can be adopted. Upon encountering the start of a file within the data, chunks can be processed incrementally. If the chunk size becomes substantial, data already processed can be written to a disk-based blob, followed by piping the rest of the data to that blob. Similar strategies can be applied as described above.

---

> I'd also like to mention that unlike busboy or @fastify/busboy, there is no way to limit the size of a file, nor the number of files or fields, nor file name length, or any other limit that these packages have. 

Well, at least you could examining the complete `content-length` to set an overall data limit.

# Alternative Solutions

There could also be a completely different approach to this as well. It involves a separate handling mechanism that absolves the developer from grappling with these complexities. If NodeJS featured a "Blob Storage" concept, it could autonomously decide whether to page `BlobParts` to disk to manage excessive data allocation. This would allow developer to create blob chunks without needing to worry about offloading data to the disk.

So you would just create many blobs chunks and the "core" would take care of automatically deciding if it should offload some data to the disc or not. and it could happen without you even knowing that it have happened.

Google provides relevant documentation on this approach [here](https://docs.google.com/presentation/d/1MOm-8kacXAon1L2tF6VthesNjXgx0fp5AP17L7XDPSM/edit#slide=id.g9060b4857_0_10) and [here](https://docs.google.com/presentation/d/11oGsu8bQHrvMicx28bB9taRTTt3DIagslM5TrwcLgAg/edit#slide=id.g8fe6c1657_0_10).

In this alternative scenario, the primary responsibility would be generating blob chunks, while the core system would manage disk paging automatically. An example implementation is shown below:

```js
async blob() {
  const chunks = []
  for await (const uint8_chunk of getStream(this)) {
    // creating many small blobs "could" automatically start paging to Disk
    const small_chunk = new Blob([ uint8_chunk ]) 
    chunks.push(small_chunk)
  }
  
  // this final blob would just hold many references to where each chunk is on the disc. 
  // so this would not hold much data or any memory, this would be fine to have in memory without problems
  return new Blob(chunks)
}
```
This approach presents advantages by globally handling the intricacies and simplifying memory and disk usage concerns for everyone by not having to think about this concerns

---

> Chrome's .formData parser is also synchronous, as it should be according to the spec,

if the parser is being synchronous or asynchronous in accordance with specifications isn't the main concern. The paramount consideration is that the parser functions as anticipated by the specifications. If different approaches yield the same outcome, the manner in which it's achieved becomes irrelevant

**Comment by KhafraDev** [comment]
First off I'd like to mention that the solutions offered either do not work in this context or are more complicated than using one of the aforementioned libraries. Chrome also works in an entirely different environment than node, where there is not a better solution other than this overly complicated approach. Node has better alternatives, even if we were to implement everything here (which I am strongly against due to how complex it is).

> the process of cleaning up files can be managed using finalizers, as exemplified in [this reference](https://github.com/node-fetch/fetch-blob/blob/a1a182e5978811407bef4ea1632b517567dda01f/from.js#L70C1-L80C2), losing references to the blob could delete the file.

There is no guarantee that those finalizers are called. If the process crashes or is killed externally, is there any guarantee that they're called? If not, do we just ignore those temporary files and let them build up? We're still facing issues with WeakRefs and finalizers in other parts of fetch.

> specifications isn't the main concern.

It *is* the main concern. The spec, for everything else, is the foremost and important thing we regard. I see no reason that it should be different for .formData, but it is.

> The paramount consideration is that the parser functions as anticipated by the specifications.

It doesn't. There are 4 WPTs that do not pass for .formData, this is an absurdly high amount of failures considering the number of tests (14) and compared to other mixin methods, which are spec-compliant.

**Comment by jimmywarting** [comment]
> This unfortunately isn't even close to a viable alternative.

I know, but this is all that you got to go on... it isn't much but at least it's something. it could work for a simple check such as not being able to upload more than 4 GiB...

>  If the process crashes or is killed externally, is there any guarantee that they're called?

There could be an alternative solution to this. 
1. create a file and open a File Descriptor (`fd`)
2. immediately unlink the `fd` 
3. any remaning `fd` can then read/write to this file as long as it's open while the file is still flagged for deletion

> If the process has opened a file and that file was flagged for deletion (using functions like `unlink()` or `remove()` in C), the file will be deleted from the filesystem when the last reference to it is closed. This is because the file's directory entry is removed as part of the deletion process.

So if the process suddenly exit or is killed then the file will be deleted afterwards.
So maybe i will refactor my own code to using that instead of `fs.promise.writeFile()`
